internal/engine/telemetrycontroller.go

```diff
@@ -0,0 +1,82 @@
+package engine
+
+import (
+	"context"
+	"os"
+	"os/exec"
+	"time"
+
+	"github.com/windmilleng/wmclient/pkg/dirs"
+
+	"github.com/windmilleng/tilt/internal/build"
+	"github.com/windmilleng/tilt/internal/engine/configs"
+	"github.com/windmilleng/tilt/internal/store"
+	"github.com/windmilleng/tilt/internal/tracer"
+	"github.com/windmilleng/tilt/pkg/model"
+)
+
+type TelemetryController struct {
+	clock build.Clock
+	// lock is a lock that is used to control access from the span exporter and the telemetry controller to the file
+	lock tracer.Locker
+	dir  *dirs.WindmillDir
+}
+
+func NewTelemetryController(lock tracer.Locker, clock build.Clock, dir *dirs.WindmillDir) *TelemetryController {
+	return &TelemetryController{
+		lock:  lock,
+		clock: clock,
+		dir:   dir,
+	}
+}
+
+func (t *TelemetryController) OnChange(ctx context.Context, st store.RStore) {
+	state := st.RLockState()
+	lastTelemetryRun := state.LastTelemetryScriptRun
+	tc := state.TelemetryCmd
+	st.RUnlockState()
+
+	if tc.Empty() || !lastTelemetryRun.Add(1*time.Hour).Before(t.clock.Now()) {
+		return
+	}
+
+	t.lock.Lock()
+	defer t.lock.Unlock()
+
+	// exec the telemetry command, passing in the contents of the file on stdin
+	cmd := exec.CommandContext(ctx, tc.Argv[0], tc.Argv[1:]...)
+	file, err := t.dir.OpenFile(tracer.OutgoingFilename, os.O_RDONLY, 0644)
+	if err != nil {
+		logError(st, err)
+		return
+	}
+	defer func() {
+		err := file.Close()
+		if err != nil {
+			logError(st, err)
+		}
+	}()
+
+	cmd.Stdin = file
```

Comment:
where does the Stdout/Stderr of this command go?

i wonder if we'll need eventually need a new kind of manifest to handle all this

Reply:
I think that's likely, but not something I wanted to do immediately.

The stdout is purposefully hidden to not distract end users with telemetry that they certainly don't care about. stderr should be surfaced if there's an error, though. Good call.